rfd/0126-backend-migrations.md

```diff
@@ -0,0 +1,258 @@
+
+## What
+
+A more reliable and robust mechanism to perform backend migrations.
+
+## Why
+
+The upgrade process requires scaling Auth down to a single instance to ensure that migrations are only performed once as well as preventing a new version and an old version from operating on the same keys with different schema. This is cumbersome and makes the upgrade process a pain point for cluster admins. Scaling Auth down and back up also results in an [uneven load](https://github.com/gravitational/teleport/issues/7029) which can cause connectivity issues and backend latency that can result in a thundering herd.
+
+Migrations to date have existed as a function that is called during Auth initialization. The migration function is left as is for a release or two and is then deleted. This makes tracking when and which migrations have been applied difficult and prevents skipping any major versions when upgrading since there are no guarantees that the latest version will include all the migrations between versions.
+
+## Details
+
+### When is a Migration required?
+
+Migrations are not needed for every change to a backend resource. Adding a new field to the resource should be a backward compatible operation by nature of Protocol Buffers and does not require an actual migration. Any major changes to a resource that alter its shape, change its encoding, or moving a single resource into several other resources are candidates for a migration.
+
+### Persistent Migrations
+
+All new migrations MUST exist in perpetuity to allow for the correct migrations to be applied regardless of which version of Teleport is being upgraded from and to. Migrations MUST be numbered and applied in sequence. The order of migrations MUST not change and a migration MUST not be altered once it has been included in a release.
+
+To make tracking and discovering migrations easier all migrations MUST be placed in `lib/auth/migrations`. Each migration should be named `$number-migration-description.go` and contain a single migration function, see the Migrations section below for details. The following is an example of four migrations:
+
+```
+./lib/auth/migrations/
+├── 0001-initial_migration.go
+├── 0002-another_migration.go
+├── 0003-some_other_migration.go
+└── 0004-yet_another_migration.go
+```
+
+### Migration History
+
+Migration history is to be stored in the backend to enable tracking which migrations have been applied, when, and whether they were successful or not.
+The key `/migrations/<migration_number>` will store the history of a migration.
+
+Cluster admins may view the migration history via `tctl migrations ls`.
+
+### Migrating Keys
+
+All migrations MUST not be performed in place, instead migrations MUST migrate both the keys and values in the backend.
+For example, to migrate the data stored in key `some/path/to/resource/<ID>` we must leave the original value at `some/path/to/resource/<ID>` as is and write the migrated value to `some/new/path/to/resource/<ID>`.
```

Comment:
Is it assumed that the entry at the old key won't be touched anymore? Are we okay with the data in the new key drifting?

Any concerns w.r.t expiration? If the new version is not heartbeating for data at the old keys, they'll eventually expire. (I think this is fine, but it means we just have to be clear that this design allows for you to run old and new auth instances side by side for a short period of time, and this is not a state we expect you to be in for very long)

Reply:
We can't let multiple instances of the Auth server run at the same time with multiple different versions, because it can create overwrite problem (when auth server old version will write the version to the old path, while the new one will write to the new path)

Reply:
@rosstimothy this sounds right to me.

Reply:
Sounds good to me too - let's go with this approach.